packages/mui-utils/src/composeClasses/composeClasses.ts

```diff
@@ -1,30 +1,48 @@
+/* eslint no-restricted-syntax: 0, prefer-template: 0, guard-for-in: 0
+   ---
+   These rules are preventing the performance optimizations below.
+ */
+
 export default function composeClasses<ClassKey extends string>(
   slots: Record<ClassKey, ReadonlyArray<string | false | undefined | null>>,
   getUtilityClass: (slot: string) => string,
   classes: Record<string, string> | undefined = undefined,
 ): Record<ClassKey, string> {
   const output: Record<ClassKey, string> = {} as any;
 
-  Object.keys(slots).forEach(
-    // `Object.keys(slots)` can't be wider than `T` because we infer `T` from `slots`.
-    // @ts-expect-error https://github.com/microsoft/TypeScript/pull/12253#issuecomment-263132208
-    (slot: ClassKey) => {
-      output[slot] = slots[slot]
-        .reduce((acc, key) => {
-          if (key) {
-            const utilityClass = getUtilityClass(key);
-            if (utilityClass !== '') {
-              acc.push(utilityClass);
-            }
-            if (classes && classes[key]) {
-              acc.push(classes[key]);
-            }
+  if (classes) {
```

Comment:
Question, does the perf improvement comes from replacing the built in array functions with for each, or from actually checking if classes [value] exists for each slot? I am asking because we had quite of code duplication in these two conditions - my only worry is that if there is a bug in the future we may miss one of the conditions.

Reply:
Most of the improvement comes from using string concatenation to build the output, likely because it avoids the temporary arrays (for `.join(' ')`). Here is a summary of how each change affected the performance (this one dates from my old PR). I've however tested again without the `if (classes)` branching change and I find no statistically significant difference, so I've removed that change and kept a single loop.

<img src="https://github.com/mui/material-ui/assets/1423607/69ea3547-9702-4378-9108-f8e5db8c4c24" width="450" />

Reply:
Cool, nice to know, thanks for the insights 💡 It's very strange that we have eslint rules that prevents us from writing more performant code 😓

Reply:
I *constantly* bump into eslint when I'm optimizing code. The neat `isObjectEmpty` function hits like 3-4 rules at once lol.